### PR TITLE
Run all tests

### PR DESCRIPTION
**Pull request by phracek**
This commit adds a feature for running all tests at once.

In the end, CI shows all results.

### CODE REVIEW

**Comment by phracek** [comment]
[test]

File: test/run [hunk 1]
```diff
@@ -195,60 +217,104 @@ test_from_dockerfile(){
...
+    # Tests with '-should-fail-' in their name should fail during a build, expecting non-zero exit status
+    check_type $RESULT "negative"
+    if [ "$?" != "0" ]; then
+      test_msg="[PASSED]"
+    else
+      test_msg="[FAILED]"
+    fi
+    printf -v test_short_summary "%s %s for '%s' %s\n" "${test_short_summary}" "$test_msg" "${app}" "run_s2i_build"
+    return 1
+  else
```
Comment (frenzymadness) [inline test/run:282]: It seems to me that these two branches contain almost exactly the same code. Would it be possible to merge them?

File: test/run [hunk 2]
```diff
@@ -12,6 +12,18 @@ declare -a WEB_APPS=({standalone,setup,setup-requirements,django,numpy,app-home,
...
+test_short_summary=''
+TESTSUITE_RESULT=0
+
+TEST_LIST="\
+test_s2i_usage
+test_docker_run_usage
+test_application
+test_application_with_user
+test_scl_variables_in_dockerfile
+test_from_dockerfile
```
Comment (frenzymadness) [inline test/run:24]: `test_from_dockerfile` uses an external repository as an application source so it does not make sense to run it more than once for each image.

File: test/run [hunk 3]
```diff
@@ -83,20 +95,30 @@ cleanup() {
...
   elif [[ "$type" == "negative" && "$result" == "0" ]]; then
     info "TEST FAILED (${type}), EXPECTED: non-zero GOT:${result}"
     cleanup
-    exit 1
+    return 1
   fi
+
+}
+check_result() {
+  # positive test & non-zero exit status = ERROR
```
Comment (frenzymadness) [inline test/run:115]: Are these comments supposed to be here or is this just copy/paste error?

File: test/run [hunk 4]
```diff
@@ -83,20 +95,30 @@ cleanup() {
   rm -rf ${test_dir}/${1}/.git
 }
 
-check_result() {
+check_type() {
```
Comment (frenzymadness) [inline test/run:98]: Could you please add a comment on what are check_[result,type] functions for? I'm confused here.

File: test/run [hunk 5]
```diff
@@ -195,60 +217,104 @@ test_from_dockerfile(){
...
+
+test_scl_variables_in_dockerfile() {
+  if [ "$OS" == "rhel7" ] || [ "$OS" == "centos7" ]; then
+     # autocleanup only enabled here as only the following tests so far use it
+     CID_FILE_DIR=$(mktemp -d)
+     ct_enable_cleanup
 
+     info "Testing variable presence during \`docker exec\`"
+     ct_check_exec_env_vars
+     check_result $?
```
Comment (frenzymadness) [inline test/run:239]: If the `check_result` here returns 1, does it terminate the function or not? And if not, will this result be reflected in the final summary?

File: test/run [hunk 6]
```diff
@@ -195,60 +217,104 @@ test_from_dockerfile(){
...
+      return 1
+    }
+  done;
+}
+
+function test_run_s2i_build() {
+  local test_msg
+  local result="$1"
+  local app="$2"
+  if [[ "$app" == *"-should-fail-"* ]]; then
```
Comment (frenzymadness) [inline test/run:272]: If an app fails to build and we expect that, all following tests should be skipped.

File: test/run [hunk 7]
```diff
@@ -12,6 +12,18 @@ declare -a WEB_APPS=({standalone,setup,setup-requirements,django,numpy,app-home,
...
+TESTSUITE_RESULT=0
+
+TEST_LIST="\
+test_s2i_usage
+test_docker_run_usage
+test_application
+test_application_with_user
+test_scl_variables_in_dockerfile
+test_from_dockerfile
+"
```
Comment (frenzymadness) [inline test/run:25]: Missing `test_application_enable_init_wrapper` here.

**Review by frenzymadness — CHANGES_REQUESTED** [review]
The first review is done. Sorry if I misunderstood some parts but the script is now kinda complex and might deserve some more comments especially for variables and functions with very similar names.

File: test/run (hunk 5 above)
Comment (phracek) [reply test/run:239]: Yes, if `check_result` returns 1, it sets a variable `TESTCASE_RESULT`, and based on the value it marks the tests as a failure.

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by frenzymadness** [comment]
I'm triaging PRs in this repository and I'll switch this one to draft to indicate it's not ready yet for a final review.

**Comment by phracek** [comment]
Sorry for a bit late response.

Let me describe this PR a bit more.
The PR adds two test suite variables: `TEST_LIST` and `TEST_VAR_DOCKER`, which contains all test cases used by s2i-python-container.
I have added the function called `check_result` which set up the `TESTCASE_RESULT` variable to True. The variable is reset before each test function in `run_all_tests` function. The function checks only the return value of the command and no other magic.

The `check_type` function was created to move this code https://github.com/sclorg/s2i-python-container/blob/master/test/run#L222.

I have created two functions that look very similar. The one is called `test_application_with_user` which `test_application` with a user and `test_application_enable_init_wrapper` which tests application with enabled wrapper.

`run_all_tests` executes all tests for the relevant applications. The main cycle overall applications remain as it is https://github.com/sclorg/s2i-python-container/blob/master/test/run#L218. `run_all_tests` is called inside it.

The code https://github.com/sclorg/s2i-python-container/blob/master/test/run#L234 , https://github.com/sclorg/s2i-python-container/blob/master/test/run#L237 and https://github.com/sclorg/s2i-python-container/blob/master/test/run#L243 were moved into functions `test_application_with_user`, `test_application_enable_init_wrapper` and  `test_scl_variables_in_dockerfile` respectivelly.

**Comment by frenzymadness** [comment]
[test-openshift]

**Comment by frenzymadness** [comment]
[test]

**Comment by frenzymadness** [comment]
I like the idea and also the output it produces! Please, rebase it.

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
@frenzymadness WDYT? Can it be merged now? ;)

**Comment by frenzymadness** [comment]
The script looks good, its output as well and the CI is green, I'm gonna merge it, good job, thank you!

**Pull request merged**